File: 1185.day-of-the-week.cpp
/*
 * @lc app=leetcode id=1185 lang=cpp
 *
 * [1185] Day of the Week
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  string dayOfTheWeek(int day, int month, int year) {
    const std::string day_of_the_week[] = {"Sunday",    "Monday",   "Tuesday",
                                           "Wednesday", "Thursday", "Friday",
                                           "Saturday"};
    constexpr int day_of_month[] = {
        0,
        31,
        31 + 28,
        31 + 28 + 31,
        31 + 28 + 31 + 30,
        31 + 28 + 31 + 30 + 31,
        31 + 28 + 31 + 30 + 31 + 30,
        31 + 28 + 31 + 30 + 31 + 30 + 31,
        31 + 28 + 31 + 30 + 31 + 30 + 31 + 31,
        31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30,
        31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30 + 31,
        31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30 + 31 + 30,
        31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30 + 31 + 30 + 31};
    int offset = 5;
    offset += day - 1;
    offset += day_of_month[month - 1] % 7;
    if (is_leap_year(year) && month > 2) {
      offset++;
    }
    for (int i = 1971; i < year; i++) {
      if (is_leap_year(i)) {
        offset += 2;
      } else {
        offset++;
      }
    }
    return day_of_the_week[offset % 7];
  }

 private:
  bool is_leap_year(int year) {
    return (year % 4 == 0) && ((year % 100 != 0) || (year % 400 == 0));
  }
};
// @lc code=end

int main() {
  Solution s;
  ASSERT_EQ(s.dayOfTheWeek(1, 1, 1971), "Friday");
  ASSERT_EQ(s.dayOfTheWeek(2, 1, 1971), "Saturday");
  ASSERT_EQ(s.dayOfTheWeek(3, 1, 1971), "Sunday");
  ASSERT_EQ(s.dayOfTheWeek(1, 2, 1971), "Monday");
  ASSERT_EQ(s.dayOfTheWeek(1, 1, 1972), "Saturday");
  ASSERT_EQ(s.dayOfTheWeek(1, 1, 1973), "Monday");
  ASSERT_EQ(s.dayOfTheWeek(31, 8, 2019), "Saturday");
  ASSERT_EQ(s.dayOfTheWeek(18, 7, 1999), "Sunday");
  ASSERT_EQ(s.dayOfTheWeek(15, 8, 1993), "Sunday");
  ASSERT_EQ(s.dayOfTheWeek(31, 8, 2100), "Tuesday");
  return 0;
}

File: 1185.day-of-the-week.rs
/*
 * @lc app=leetcode id=1185 lang=rust
 *
 * [1185] Day of the Week
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn day_of_the_week(d: i32, mut m: i32, mut y: i32) -> String {
        let days = [
            "Sunday",
            "Monday",
            "Tuesday",
            "Wednesday",
            "Thursday",
            "Friday",
            "Saturday",
        ];
        if m < 3 {
            m += 12;
            y -= 1;
        }
        let c = y / 100;
        y %= 100;
        let w = (c / 4 - 2 * c + y + y / 4 + 13 * (m + 1) / 5 + d - 1) % 7;
        days[((w + 7) % 7) as usize].to_string()
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::day_of_the_week(31, 8, 2019),
        "Saturday".to_string()
    );
}

#[test]
fn test2() {
    assert_eq!(Solution::day_of_the_week(18, 7, 1999), "Sunday".to_string());
}

#[test]
fn test3() {
    assert_eq!(Solution::day_of_the_week(15, 8, 1993), "Sunday".to_string());
}
